pydarn/pydmap/io.py

```diff
@@ -0,0 +1,1296 @@
+
+Exceptions:
+-----------
+EmptyFileError
+CursorError
+DmapDataError
+DmapDataTypeError
+ZeroByteError
+MismatchByteError
+ValueError
+
+Future work
+-----------
+logger improvement
+    time performance can improve if we turn off the logger
+DmapWrite cleanup
+    updating and improving DmapWrite
+Date stream reading/writing
+    add the ability to read/write from data streams <-- this may be working but
+                                                        not tested well
+Organization
+    rethink public and private methods? <--- discussion
+
+Notes
+-----
+DmapRead and DmapWrite are updated and improved versions of
+backscatter's pydmap:
+git@github.com:SuperDARNCanada/backscatter.git
+written by Keith Kotyk
+"""
+
+import os
+import struct
+import numpy as np
+import logging
+import collections
+
+from pydarn import pydmap_exceptions
+
+from pydarn import DmapArray
+from pydarn import DmapScalar
+
+# Keeping these global definitions for readability purposes
+# Data types use in DMAP files
+DMAP = 0
+CHAR = 1
+SHORT = 2
+INT = 3
+FLOAT = 4
+DOUBLE = 8
+STRING = 9
+LONG = 10
+UCHAR = 16
+USHORT = 17
+UINT = 18
+ULONG = 19
+
+# Dictionary of DMAP types (key) to quickly convert the format and byte size
+# (value-tuple)
+DMAP_DATA_TYPES = {DMAP: ('', 0),
+                   CHAR: ('c', 1),
+                   SHORT: ('h', 2),
+                   INT: ('i', 4),
+                   FLOAT: ('f', 4),
+                   DOUBLE: ('d', 8),
+                   STRING: ('s', 1),
+                   LONG: ('q', 8),
+                   UCHAR: ('B', 1),
+                   USHORT: ('H', 2),
+                   UINT: ('I', 4),
+                   ULONG: ('Q', 8)}
+
+pydarn_logger = logging.getLogger('pydarn')
+
+class DmapRead():
+    """
+    Reading and testing the integrity of DMAP files/stream.
+    ...
+
+    Attributes:
+    ----------
+    dmap_file : str
+        DMAP file/stream name/data
+    cursor : int
+        Current position in the byte array
+    dmap_end_bytes : int
+        The length of the byte array
+
+    Methods:
+    --------
+    test_initial_data_integrity()
+        Quickly reads the byte array for any errors
+    read_records()
+        Reads the byte array to obtain the DMAP records
+    zero_check(element, element_name)
+        Checks if the element is equal to zero
+    check_data_type(data_type, data_name)
+        Checks if the data_type exists in DMAP_DATA_TYPE dictionary
+    read_record()
+        Reads a record from the byte array
+    read_scalar()
+        Reads a scalar from the byte array
+    read_array(record_size)
+        Reads an array from the byte array
+    read_string_DMAP_array(Shape, array_type)
+        Reads an array of strings/DMAP types into an array
+    read_data()
+        Reads the data type from the byte array
+    read_numerical_array(data_type, shape, total_number_cells)
+        Reads an array into a numpy array
+    """
+
+    def __init__(self, dmap_file: str, data_stream=False):
```

Comment:
Not in 3.6 :) 

I use dmap_buffer later :/ ... I personally don't mind `file` since the main use case is to use a file.

Reply:
Maybe dmap_bytes? I know that file is the main use case and that does make sense, but it doesn't completely add up with the doc string and its not 100 clear to those who do work want to work with realtime streams. Maybe a clarification in the doc string then that file can represent any stream of bytes whether it be file or network?

Reply:
`dmap_bytes` still makes no sense since it doesn't represent a file correctly. 

I didn't know I was making something for real-time streams thought this was a DaVitpy replacement which was not used in real-time streams :p 

I will update the documentation until a better suggestion comes along.